integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+    private static Association prepare(Association a) {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(Association.class.getName(), "prepare", a);
+        }
+        assert a != null; // already vetted
+        assert !a.suspended(); // can't be in T2
+        try {
+            if (a.connection().isReadOnly()) {
+                a = a.reset();
+                assert a.branchState() == Association.BranchState.NON_EXISTENT_TRANSACTION;
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    LOGGER.logp(Level.FINE, Association.class.getName(), "prepare",
+                                "Removing association {0}", a);
+                }
+                // Critically important: remove the association.
+                a = null;
+            }
+        } catch (SQLException e) {
+            throw new UncheckedSQLException(e);
+        }
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(Association.class.getName(), "preapre", a);
+        }
+        return a;
+    }
+
+    // (UnaryOperator for supplying via method reference to remap() above.)
+    private static Association forgetAndReset(Association a) {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(Association.class.getName(), "forgetAndReset", a);
+        }
+        assert a != null; // already vetted
+        try {
+            a = a.forgetAndReset();
+        } catch (SQLException e) {
+            throw new UncheckedSQLException(e);
+        }
+        assert a.branchState() == Association.BranchState.NON_EXISTENT_TRANSACTION;
+        if (LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.logp(Level.FINE, Association.class.getName(), "forgetAndReset",
+                        "Removing association {0}", a);
+        }
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(Association.class.getName(), "forgetAndReset", "null");
+        }
+        // Critically important: remove the association.
+        return null;
+    }
+
+    private static String flagsToString(int flags) {
+        switch (flags) {
+        case TMENDRSCAN:
+            return "TMENDRSCAN (" + flags + ")";
+        case TMFAIL:
+            return "TMFAIL (" + flags + ")";
+        case TMJOIN:
+            return "TMJOIN (" + flags + ")";
+        case TMNOFLAGS:
+            return "TMNOFLAGS (" + flags + ")";
+        case TMONEPHASE:
+            return "TMONEPHASE (" + flags + ")";
+        case TMRESUME:
+            return "TMRESUME (" + flags + ")";
+        case TMSTARTRSCAN:
+            return "TMSTARTRSCAN (" + flags + ")";
+        case TMSUCCESS:
+            return "TMSUCCESS (" + flags + ")";
+        case TMSUSPEND:
+            return "TMSUSPEND (" + flags + ")";
+        default:
+            return String.valueOf(flags);
+        }
+    }
+
+
+    /*
+     * Inner and nested classes.
+     */
+
+
+    static record Association(BranchState branchState,
```

Comment:
Holds a `Connection` together with a branch association state and information needed to restore its initial state.  Associations are immutable and logically move through a state machine defined by the XA specification.